library/core/src/slice/mod.rs

```diff
@@ -2692,7 +2692,7 @@ impl<T> [T] {
     /// This reordering has the additional property that any value at position `i < index` will be
     /// less than or equal to any value at a position `j > index`. Additionally, this reordering is
     /// unstable (i.e. any number of equal elements may end up at position `index`), in-place
-    /// (i.e. does not allocate), and *O*(*n*) worst-case. This function is also/ known as "kth
+    /// (i.e. does not allocate), and *O*(*n*^2) worst-case. This function is also/ known as "kth
```

Comment:
Are you sure the current implementation is actually O(n²) worse-case? It looks to me like it's doing a bunch of work in pivot selection to avoid being quadratic:

https://github.com/rust-lang/rust/blob/481725984b4cd94ef5c00917b01c1771b6e5299c/library/core/src/slice/sort.rs#L692-L716

At least I think that any documentation update here should emphasize that this is average-case O(n), since that's the point of the function existing.

Reply:
I'm relying on the statement that this is O(n²) worst-case from here: https://github.com/rust-lang/rust/issues/102451

Reply:
I noted it. I’m not sure if `select_nth_unstable_by_key` is also affected.

Reply:
Perhaps a better option would be to rewrite the algorithm based on one of the methods linked in the issue.

Reply:
Maybe as a first step here do the simple change to update the documentation just to change "worst-case" to "average"?  Since the issue shows it's currently incorrect.

Then a follow-up PR could do the same [fallback to heapsort](https://rust-lang.zulipchat.com/#narrow/stream/122652-new-members/topic/Hello.20from.20Anselm/near/321697132) that `sort_unstable` does to avoid being worse-cast O(n²).